pageserver/src/layered_repository/inmemory_layer.rs

```diff
@@ -0,0 +1,519 @@
+    fn put_page_version(&self, blknum: u32, lsn: Lsn, pv: PageVersion) -> Result<()> {
+        trace!(
+            "put_page_version blk {} of {} at {}/{}",
+            blknum,
+            self.rel,
+            self.timelineid,
+            lsn
+        );
+        let mut inner = self.inner.lock().unwrap();
+
+        let old = inner.page_versions.insert((blknum, lsn), pv);
+
+        if old.is_some() {
+            // We already had an entry for this LSN. That's odd..
+            warn!(
+                "Page version of rel {:?} blk {} at {} already exists",
+                self.rel, blknum, lsn
+            );
+        }
+
+        // Also update the relation size, if this extended the relation.
+        if self.rel.is_blocky() {
+            let mut iter = inner.relsizes.range((Included(&Lsn(0)), Included(&lsn)));
+
+            let oldsize;
+            if let Some((_entry_lsn, entry)) = iter.next_back() {
+                oldsize = *entry;
+            } else {
+                oldsize = 0;
+                //bail!("No old size found for {} at {}", self.tag, lsn);
+            }
+            if blknum >= oldsize {
+                trace!(
+                    "enlarging relation {} from {} to {} blocks at {}",
+                    self.rel,
+                    oldsize,
+                    blknum + 1,
+                    lsn
+                );
+                inner.relsizes.insert(lsn, blknum + 1);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Remember that the relation was truncated at given LSN
+    fn put_truncation(&self, lsn: Lsn, relsize: u32) -> anyhow::Result<()> {
+        let mut inner = self.inner.lock().unwrap();
+        let old = inner.relsizes.insert(lsn, relsize);
+
+        if old.is_some() {
+            // We already had an entry for this LSN. That's odd..
+            warn!("Inserting truncation, but had an entry for the LSN already");
+        }
+
+        Ok(())
+    }
+
+    /// Remember that the relation was truncated at given LSN
+    fn put_unlink(&self, lsn: Lsn) -> anyhow::Result<()> {
+        let mut inner = self.inner.lock().unwrap();
+
+        assert!(inner.drop_lsn == Lsn(0));
+        inner.drop_lsn = lsn;
+
+        info!("dropped relation {} at {}", self.rel, lsn);
+
+        Ok(())
+    }
+
+    ///
+    /// Write the this in-memory layer to disk, as a snapshot layer.
+    ///
+    /// The cutoff point for the layer that's written to disk is 'end_lsn'.
+    ///
+    /// Returns new layers that replace this one. Always returns a
+    /// SnapshotLayer containing the page versions that were written to disk,
+    /// but if there were page versions newer than 'end_lsn', also return a new
+    /// in-memory layer containing those page versions. The caller replaces
+    /// this layer with the returned layers in the layer map.
+    ///
+    fn freeze(
+        &self,
+        end_lsn: Lsn,
+        walredo_mgr: &dyn WalRedoManager,
+    ) -> Result<Vec<Arc<dyn Layer>>> {
```

Comment:
Hmm. What is the "latest LSN"? The individual layers don't track such a thing, the closest we have is the last_valid_lsn in LayeredTimeline. And that's what the caller passes here as the end_lsn argument.

One problem is that one WAL record can modify many pages. All the modifications are parsed in `save_decoded_record', which turns one WAL record into a sequence of `put` calls in the Repository. For example, a page split record could result in a call sequence like this:
```
  timeline.advance_last_record_lsn(90) // previous record

  timeline.put_wal_record(rel, 10 /* blk */, 100 /* LSN */);  // left page
  timeline.put_wal_record(rel, 11 /* blk */, 100 /* LSN */);  // right page
  timeline.advance_last_record_lsn(100)
```
The `advance_last_record` calls work like atomic boundaries. If a checkpoint happens in between the two put_wal_record() calls and we need to freeze the layer, we would freeze and create the on-disk layer as of LSN 90 and keep the changes for LSN 100 in the new in-memory layer.

Reply:
I'm not sure if I understand.

I think that layer mutation and snapshotting should be atomic relative to one another. I would much prefer a design that will only perform a snapshot when the layers are in a consistent state.

Then the snapshot/write-file process goes like this:

start:

    LAYER3(memory) <- all writes go here.
    LAYER2(file)
    LAYER1(file)

step 2: atomic relative to wal record processing, so no "torn" writes are possible.

    LAYER4(memory) <- all writes now land here
    LAYER3(memory) <- now read-only; this is a temporary state until step 3 finishes.
    LAYER2(file)
    LAYER1(file)

step 3: write out the previous top layer to a file; atomically swap the file `LAYER3'` for the in-memory `LAYER3`.

    LAYER4(memory) <- all writes now land here
    LAYER3'(file)
    LAYER2(file)
    LAYER1(file)

The nice properties of this are:
- Creating a snapshot file doesn't involve splitting the layer into "new" and "old" layers, it's just writing out 100% of the data in that layer.
- Creating a snapshot file doesn't need to copy any `BtreeMap` data structures. Because the map is now read-only, we can just iterate over its contents, streaming its data into a file (with very little memory overhead).
- All of the steps are either fast/atomic (create new top layer) or don't modify the in-memory layer (writing the file), so there is no concurrency with writers. Writers only modify the top layer, while snapshots only deal with layers that are static & unchanging. This makes it a lot easier to ensure correctness because there is minimal opportunity for race conditions.
- The previous layers and the new layers have a 1:1 relationship, which makes debugging simple.
- `LAYER3` and `LAYER3'` contain the same data, so it's really simple to demonstrate that the code is correct.

Reply:
+1 on that approach.

A simple way to make step 2 atomic with WAL processing would be to have a global mutex that you hold when you process change for one WAL records, and also in step 2 above. A global mutex seems pretty heavy-weight, but if it's only those two threads that compete for it, and "step 2" is pretty fast, I guess it would be ok.

I'll give that a try.